File: src/main/java/cn/wjybxx/dson/text/DsonToken.java
/*
 * Copyright 2023 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.dson.text;

import java.util.Objects;

/**
 * token可能记录位置更有助于排查问题
 *
 * @author wjybxx
 * date - 2023/6/2
 */
public class DsonToken {

    private final TokenType type;
    private final Object value;
    private final int pos;

    /**
     * @param pos token所在的位置，-1表示动态生成的token
     */
    public DsonToken(TokenType type, Object value, int pos) {
        this.type = Objects.requireNonNull(type);
        this.value = value;
        this.pos = pos;
    }

    public TokenType getType() {
        return type;
    }

    public Object getValue() {
        return value;
    }

    public int getPos() {
        return pos;
    }

    //

    public String castAsString() {
        return (String) value;
    }

    public char firstChar() {
        String value = (String) this.value;
        return value.charAt(0);
    }

    public char lastChar() {
        String value = (String) this.value;
        return value.charAt(value.length() - 1);
    }

    //

    public boolean fullEquals(DsonToken dsonToken) {
        if (this == dsonToken) {
            return true;
        }
        if (pos != dsonToken.pos) return false;
        if (type != dsonToken.type) return false;
        return Objects.equals(value, dsonToken.value);
    }

    /**
     * 默认忽略pos的差异 -- 通常是由于换行符的问题。
     * 由于动态生成的token的pos都是-1，因此即使比较pos，对于动态生成的token之间也是无意义的，
     * 既然动态生成的token之间的相等性是忽略了pos的，那么正常的token也需要忽略pos。
     */
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        DsonToken dsonToken = (DsonToken) o;

        if (type != dsonToken.type) return false;
        return Objects.equals(value, dsonToken.value);
    }

    @Override
    public int hashCode() {
        int result = type.hashCode();
        result = 31 * result + (value != null ? value.hashCode() : 0);
        return result;
    }

    @Override
    public String toString() {
        return "DsonToken{" +
                "type=" + type +
                ", value=" + value +
                ", pos=" + pos +
                '}';
    }
}

File: csharp/Dson/Text/DsonToken.cs
/*
 * Copyright 2023 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Dson.Text;

public class DsonToken : IEquatable<DsonToken>
{
#nullable disable
    public readonly DsonTokenType Type;
    public readonly object Value;
    public readonly int Pos;
#nullable enable
    /// <summary>
    /// 
    /// </summary>
    /// <param name="type">token的类型</param>
    /// <param name="value">token关联的值</param>
    /// <param name="pos">pos token所在的位置，-1表示动态生成的token</param>
    public DsonToken(DsonTokenType type, object? value, int pos) {
        this.Type = type;
        this.Value = value;
        this.Pos = pos;
    }

    public string castAsString() {
        return (string)Value!;
    }

    public char firstChar() {
        String value = (String)this.Value!;
        return value[0];
    }

    public char lastChar() {
        string value = (string)this.Value!;
        return value[value.Length - 1];
    }

    #region equals

    // Equals默认不比较位置

    public bool Equals(DsonToken? other) {
        if (ReferenceEquals(null, other)) return false;
        if (ReferenceEquals(this, other)) return true;
        return Type == other.Type && Equals(Value, other.Value);
    }

    public override bool Equals(object? obj) {
        if (ReferenceEquals(null, obj)) return false;
        if (ReferenceEquals(this, obj)) return true;
        if (obj.GetType() != this.GetType()) return false;
        return Equals((DsonToken)obj);
    }

    public override int GetHashCode() {
        return HashCode.Combine((int)Type, Value);
    }

    public static bool operator ==(DsonToken? left, DsonToken? right) {
        return Equals(left, right);
    }

    public static bool operator !=(DsonToken? left, DsonToken? right) {
        return !Equals(left, right);
    }

    #endregion

    public override string ToString() {
        return $"{nameof(Type)}: {Type}, {nameof(Value)}: {Value}, {nameof(Pos)}: {Pos}";
    }
}